draft-specifications/Model.md

```diff
@@ -306,21 +386,26 @@ value or rejected with the promises's eventual rejection reason.
 >
 > Tracking https://github.com/ocapn/ocapn/issues/55
 
-## Error
+OCapN does not maintain [Pass Invariant Equality](#pass-invariant-equality) for
+promises.
 
-A value capturing the reason for rejecting a delivery.
+# Pass Invariant
+
+> Any property that holds both locally and remotely between a pair of OCapN peers
+> during a session is a Pass Invariant.
+> For example, for some value types, OCapN maintains invariant that any pair of
+> values of that type, if and only if those values are Equal locally, they will
+> be equal remotely.
+> Furthermore, if the remote peer sends one value back, it will remain equal to
+> the value retained locally.
+
+## Pass Invariant Equality
+
+A type holds Equality invariant over passage between OCapN peers if values of that
+type can be passed from a local peer to a remote peer, then returned to the local peer,
+and the value that was sent is Equal to the value received.
+
+> The operators necessary for checking Equality may vary for each language and even
+> for types in one language.
+> For example, `Object.is`
```

Comment:
Actually, you should use `Object.is` as a clarifying counter-example. Containers do ***NOT*** have pass invariant `Object.is` equality.

Reply:
I began the adventure of writing in pass-invariant equality for every model type and quickly mired in details and exceptions that are outside consensus. We can’t yet say that Errors (which are Containers) are not pass-invariant by `Object.is` (lacking consensus about errors, which either need to be or be associated with a capability) and we also can’t say that ByteArrays (Atoms) are pass-invariant by `Object.is`. I believe we need a new issue on Pass-Invariant Equality and to fold that detail in follow-up. To that end, I can leave Pass-Invariant Equality for References or pull it out for a later holistic treatment.

Reply:
If we cannot settle this here, then we should postpone to a later PR. But I believe we've already committed to clear answers to all of these, with the possible exception of errors, and I'm happy with those answers:

- The pass-invariant ocapn equality of ***Targets*** is correctly reflected in JS as `Object.is`

- The pass-invariant ocapn equality of ***Atoms*** that are represented in JS by ***JS primitives*** is correctly reflected in JS as `Object.is`.

- The pass-invariant ocapn equality of ***ByteArrays***, because they are represented in JS as non-JS primitives that we are *not* going to intern, cannot be reflected in JS as `Object.is`. Note that this awkward difference in the JS behavior associated with ocapn Strings vs ocapn ByteArrays affects only the JS programmer. No such asymmetry bleeds into the ocapn level.

- The pass-invariant ocapn equality of ***Containers*** --- Lists, Structs, Taggeds --- is most definitely not reflected in JS as `Object.is`.

- OCapN ***Promises*** have no pass-invariant ocapn equality, so the question of `Object.js` applied to the JS representation of an ocapn promises does not arise. Other pass-invariant aspects of promises, like that they remain promises, and whatever we specify wrt observable settlement, are not *equality* invariants.

IMO, the only hard one:

- OCapN ***Errors*** have no pass-invariant ocapn equality, so the question of `Object.js` applied to the JS representation of an ocapn errors does not arise. Other pass-invariant aspects of errors, like that they remain errors, and that at least some of their properties (like `name` but TBD) are considered semantic and must round trip, are not *equality* invariants.

The reason the Error case is so problematic is that, for diagnostic purposes only, they carry an associated identity (variously referred to an an `errorId`) for looking up associated out-of-band diagnostic info. The naming of this identity will somehow need to be reflected in the ocapn wire protocol. It remains to be seen whether we succeed at specifying this while maintaining a clear separation between semantics and diagnostics. If this separation collapses, errors may indeed come to have cap-like unforgeable identity. But I really hope we can keep this distinction, and IMO we should proceed to try because we may succeed.

But, because of the remaining uncertainty about errors, perhaps this part alone should be postponed to a later PR.

Reply:
Having written out the above response, I'm wondering about this new classification of Errors as Containers as well. For now, until we understand errors better, what do you think of just having them be their own distinct category outside of Atom, Container, Reference/Capability?

Reply:
I’ll tentatively move Error into its own category, pending consensus on whether it is a container or reference.